include/cxxeffect/task.hpp

```diff
@@ -0,0 +1,53 @@
+#pragma once
+#include <functional>
+#include <future>
+
+namespace eff {
+    template <class F, class T>
+    using flatmap_t = decltype(std::declval<F>()(std::declval<T>()).unsafeRunSync());
```

Comment:
Let's introduce a type variable for this, called such like `return_t`, or in the same way that c++std might do it. Depending on the return type of an impure method makes my FP-spine tingle.

Reply:
Would you be alright if we did it using the C++ standard library? This is what the updated code will look like, and we'll be able to remove both `map_t` and `flatmap_t`!

![image](https://user-images.githubusercontent.com/36810712/128124465-69a11430-0e4c-4272-b46d-0dd296169b7c.png)

Reply:
In one of my stackoverflow posts someone mentioned `invoke_result_t` and I could never figure out how to use it. I'm worried about in `flatMap` if `invoke_result_t` doesn't conform to `task<*>` (also, should still be named `result_t` instead of `type`).

Reply:
If you'd like, we could take advantage of C++20 Concepts in order to require that F produces a task type. It'll give a compile-time error if someone tries to pass in a function that doesn't produce a task

Reply:
This is what that would look like:
```cpp
// This is assuming the existence of a concept named task_producer
// Which checks that a type produces a task when taking an input of type T
auto flatMap(task_producer<T> func) const {
    return {
        [f = std::move(func), ta = *this]() {
            return f(ta.unsafeRunSync()).unsafeRunSync();
        }
    };
}
```

Reply:
Ah yeah. A concept would work. I might call it a `task_bind` though. I'll check with the cats-effect people.